channeldb/payment_control_test.go

```diff
@@ -768,6 +768,214 @@ func TestPaymentControlDeletePayments(t *testing.T) {
+	// template to create the corresponding test payments in the database.
+	// Note: The payment id and number of htlc attempts of each payment will
+	// be added to this slice when creating the payments below.
+	// This allows the slice to be used directly for testing purposes.
+	payments := []payment{
+		{status: StatusFailed},
+		{status: StatusFailed},
+		{status: StatusSucceeded},
+		{status: StatusInFlight},
+	}
+
+	attemptID := uint64(0)
+
+	for i := 0; i < len(payments); i++ {
+		info, attempt, preimg, err := genInfo()
+		require.NoError(t, err, "unable to generate htlc message")
+
+		// Set the payment id accordingly in the payments slice.
+		payments[i].id = info.PaymentIdentifier
+
+		attempt.AttemptID = attemptID
+		attemptID++
+
+		// Init the payment.
+		err = pControl.InitPayment(info.PaymentIdentifier, info)
+		require.NoError(t, err, "unable to send htlc message")
+
+		// Register and fail the first attempt for all payments.
+		_, err = pControl.RegisterAttempt(info.PaymentIdentifier,
+			attempt)
+		require.NoError(t, err, "unable to send htlc message")
+
+		htlcFailure := HTLCFailUnreadable
+		_, err = pControl.FailAttempt(
+			info.PaymentIdentifier, attempt.AttemptID,
+			&HTLCFailInfo{
+				Reason: htlcFailure,
+			},
+		)
+		require.NoError(t, err, "unable to fail htlc")
+
+		// Increase the HTLC counter in the payments slice for the
+		// failed attempt.
+		payments[i].htlcs++
+
+		// Depending on the test case, fail or succeed the next
+		// attempt.
+		attempt.AttemptID = attemptID
+		attemptID++
+
+		_, err = pControl.RegisterAttempt(info.PaymentIdentifier,
+			attempt)
+		require.NoError(t, err, "unable to send htlc message")
+
+		switch payments[i].status {
+
+		// Fail the attempt and the payment overall.
+		case StatusFailed:
+			htlcFailure := HTLCFailUnreadable
+			_, err = pControl.FailAttempt(
+				info.PaymentIdentifier, attempt.AttemptID,
+				&HTLCFailInfo{
+					Reason: htlcFailure,
+				},
+			)
+			require.NoError(t, err, "unable to fail htlc")
+
+			failReason := FailureReasonNoRoute
+			_, err = pControl.Fail(info.PaymentIdentifier,
+				failReason)
+			require.NoError(t, err, "unable to fail payment hash")
+
+		// Settle the attempt
+		case StatusSucceeded:
+			_, err := pControl.SettleAttempt(
+				info.PaymentIdentifier, attempt.AttemptID,
+				&HTLCSettleInfo{
+					Preimage: preimg,
+				},
+			)
+			require.NoError(t, err, "no error should have been "+
+				"received from settling a htlc attempt")
+
+		// We leave the attmpet in-flight by doing nothing.
+		case StatusInFlight:
+		}
+
+		// Increase the HTLC counter in the payments slice for the
+		// attempt above depending on the final status of the payment.
+		payments[i].htlcs++
+	}
+
+	// Given indices corresponding to the payments slice created above this
+	// function asserts that exactly the same payments exist when fetching
+	// payments from the database.
+	assertPayments := func(paymentsIdxs ...int) {
```

Comment:
Just to be sure, do you prefer the function `assertPayments` to live outside together with some of the other assert helper function, and take the payments structure directly as argument instead of the indices (which could be of type []*MPPayment instead)?

Reply:
Yeah, was thinking a separate function rather than inlining it, can still use indexes:
```
func assertPaymentsPresent(t * testing.T, [] payment) {
    dbPayments, err := db.FetchPayments()
    ...
}
```

Just cuts the need for the `Outer` loop, and can potentially be reused one day.

Reply:
I have refactored the testing by creating a helper function `assertPayments` and also a helper function `createTestPayments` that registers those payments and populates the data into the `payments` slice: 1e613b6bbcf9d8b19c0a5a8d7e3a725486296242

Since these function can also directly be used the the test case for `DeleteAllPayments` (from which I took a lot of inspiration for the testing code!), I also refactored it to make use of the helper functions. 

I did create its own commit without squashing for easier code review. Please let me know if there is something you feel is missing or you have another idea for. Thanks!

Reply:
very nice! Thanks for taking the time on this :D Makes for very nice, readable tests.

Reply:
Cool! I will then squash it and also address your nits!